Allow version specifiers for pip install

HD650 opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
Allow version specifiers for pip install.
Finish #26299, #19690
<!--- If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request

##### COMPONENT NAME
<!--- Name of the module, plugin, module or task -->
pip.py
##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
ansible 2.7.0.dev0 (pip-version 0adb84aab0) last updated 2018/06/21 09:49:21 (GMT -400)
  config file = None
  configured module search path = [u'/home/zhikangzhang/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/zhikangzhang/Desktop/ansible/lib/ansible
  executable location = /home/zhikangzhang/Desktop/ansible/bin/ansible
  python version = 2.7.14 (default, Mar 14 2018, 16:45:33) [GCC 8.0.1 20180222 (Red Hat 8.0.1-0.16)]
```


##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful. -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
support version specifiers, also support check mode
```
- name: install package with version specifiers
  pip:
    name: "some_package"
    version: "<100.0.0,!=1.0.0,>0.0.0"
- name: check mod
  pip:
    name: "some_package_again"
    version: "<100.0.0,!=1.0.0,>0.0.0"
  check_mode: yes
- name: also support install multi packages with one name string
  pip:
    name: "package_one>=1.11.1,<=1.12.1,package_two,package_three<=2.1.1,!=1.1.1"
- name: or install multi packages with list of name
  pip:
    name:
      - "package_one>=1.11.1,<=1.12.1"
      - "package_two"
      - "package_three<=2.1.1,!=1.1.1"
```

Commit:
Allow version specifiers for pip

Commit:
Add integration test for version specifiers

Commit:
Handle exception that version number is invalid

samdoran commented:
cc @abadger

webknjaz commented on test/integration/targets/pip/tasks/pip.yml:
```diff
@@ -273,3 +273,56 @@
+- name: test the check_mod
+  pip:
+    name: "{{ pip_test_package }}"
+    version: "<100.0.0,!=1.0.0,>0.0.0"
```
Please also check `>1,<1` and `>1,==1`

abadger commented:
Implement the param as a list rather than a string.  Make sure you then have a test that both version: "single version string" and version: [ list, of, version, specs] work

HD650 commented:
> Implement the param as a list rather than a string. Make sure you then have a test that both version: "single version string" and version: [ list, of, version, specs] work
@abadger do you mean one element of the version list corresponds to one element of the name list like this:
```
pip:
  name: 
    - package one
    - package two
  version:
    - version specifier for pk one
    - version specifier for pk two
```
or just convert string version specifier to list like ```>=1.0.1,!=1.0.0,<2.0.0``` to ```[>=1.0.1, !=1.0.0, <2.0.0]```

abadger commented:
Associating names to versions by list index isn't good UI so we shouldn't
do it that way.

 Other packaging modules place the version strings into the names and then
parse those apart.

I'm not sure, is that the best way to do this (deprecating the separate
version parameter) or would it be better to make it an error for version to
be specified together with more than one name?

HD650 commented:
@abadger 
Could we still use version parameter for the case that user provides one name? And if the user wants to install multi packages in one task, they have to place version specifier in names or we raise an error. 
And I think version parameter should not accept lists, this will lead people to think that they can associate names to versions by list index.

abadger commented:
Yes

On Thu, Jun 21, 2018, 10:52 AM Zhikang Zhang <notifications@github.com>
wrote:

> @abadger <https://github.com/abadger>
> Could we still use version parameter for the case that user provides one
> name? And if the user wants to install multi packages in one task, they
> have to place version specifier in names or we raise an error.
> And I think version parameter should not accept lists, this will lead
> people to think that they can associate names to versions by list index.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ansible/ansible/pull/41792#issuecomment-399189562>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAMxWs4ZtDuchYDn5W7oM_VAidXlZvHVks5t-91SgaJpZM4UyHwi>
> .
>

Commit:
Fix the version parse problem

robinro commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -271,15 +273,22 @@ def _get_packages(module, pip, chdir):
     '''Return whether or not package is installed.'''
+    if version is not None:
+        if version[0].isdigit():
+            version = "==%s" % version
```
This is repeated from above and could be extracted to a function or called earlier..

HD650 replied:
ok.

robinro commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -271,15 +273,22 @@ def _get_packages(module, pip, chdir):
+        try:
+            vp = VersionPredicate("%s (%s)" % (name, version))
+        except ValueError:
+            module.fail_json(msg="Can parse invalid version number %s" % version)
```
Are there any packages which use abnormale versioning schemes and could be broken here?

HD650 replied:
Just in case user input some typo version here.

webknjaz replied:
"Can" or "Cannot"?

robinro commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -271,15 +273,22 @@ def _get_packages(module, pip, chdir):
             continue
 
-        if pkg_name == name and (version is None or version == pkg_version):
+        if (pkg_name.lower() == name.lower()) and (version is None or vp.satisfied_by(pkg_version)):
```
Why is the `.lower()` needed here?

HD650 replied:
A corner case I found in my machine like user want to check `django`, but the ```pip list``` return ```Django```. Pip is case insensitive but some package in ```pip list``` have capital name.

webknjaz replied:
If you'll have a class for distribution, you'll be able to encapsulate lot's of implementation details there and even keep a normalized package name there internally.

robinro commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -250,7 +250,9 @@ def _get_full_name(name, version=None):
-        resp = name + '==' + version
+        if version[0].isdigit():
+            version = "==%s" % version
+        resp = "\"%s (%s)\"" % (name, version)
```
This works for new installs? What is this new format used for?

HD650 replied:
I'm not sure whether this is compatible with all pip version... I will have more tests.

webknjaz replied:
FTR It's incompatible with outdated versions of pip

robinro reviewed (commented):
@HD650 Thanks for the PR!
VersionPredicate looks like the right thing to use.

Can you
* Implement version parsing if given as part of the name
* Add tests for multiple versions given via the name (the same way requirements.txt allows: https://pip.pypa.io/en/stable/user_guide/#requirements-files)

Commit:
Fix pip install compatible issue

Commit:
Enable version specifiers in names

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -222,19 +222,34 @@
 from ansible.module_utils._text import to_native
 from ansible.module_utils.six import PY3
 
-
```
Why did you need to remove this line?

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -222,19 +222,34 @@
+def _version_number_to_formula(version):
+    if version[0].isdigit():
+        return "==%s" % version
+    else:
```
You can just `return` without `else`.

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -250,7 +265,7 @@ def _get_full_name(name, version=None):
         resp = name
     else:
-        resp = name + '==' + version
+        resp = "\"%s %s\"" % (name, _version_number_to_formula(version))
```
This syntax with parens isn't supported in outdated pip

robinro replied:
Why change the old format at all? That works on all relevant pip versions and also allows multiple versions joined by comma.

HD650 replied:
@robinro Sorry I'm not sure which old format you mentioned. 
If you mean this one `name + '==' + version`, I think it doesn't support the case `version` is a formula like `>=1,11,1,<=1.12.1`. Here the new format make package and version formula become like `"django ==1.11.1"` or `"django >=1.11.1,<=1.12.1,!=1.11.3"` which can be parsed by pip. 
If you mean the old format like this `"\"%s (%s)\"" % (name, version)`, I remove the brackets because they are unneccessary.

webknjaz replied:
It's probably because of how distutils' predicate parser work, that's why we need to have a good distribution requirement representation.

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -523,6 +552,10 @@ def main():
 
         if name:
             for pkg in name:
+                pkg_name, version_formula = _parse_name_str(pkg)
```
Instead of complicating logic here, I'd implement a class representing a package with its version constraints and `__str__()` method for rendering.

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -557,7 +590,11 @@ def main():
 
                 for pkg in name:
-                    is_present = _is_present(pkg, version, pkg_list, pkg_cmd)
+                    pkg_name, version_formula = _parse_name_str(pkg)
```
This code basically repeats one above, which supports my previous point that having a separate class for distribution entity would be a good abstraction to have.

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -509,6 +529,15 @@ def main():
 
+        # check invalid combination of arguments
+        if version is not None:
+            if len(name) > 1:
```
This check is invalid because strings are iterable sequences.
`len('CherryPy') == 8`, `len(['aiohttp', 'aiohttp_jinja2']) == 2`.

Please use some of:
```python
from ansible.module_utils.common._collections_compat import is_sequence
...
is_sequence(..., include_strings=False)
```

or

```python
from ansible.module_utils.common._collections_compat import is_string
...
not is_string(...)
```

HD650 replied:
I think the module guarantee that name must be a list. But yes just `len(name)` is ambiguous and it's better to use `is_sequence`.

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -509,6 +529,15 @@ def main():
+                module.fail_json(msg="Version arg is not available for installing multi-packages.")
+            else:
+                pkg_name, version_formula = _parse_name_str(name[0])
+                if version_formula is not None:
```
Returning "magic" status variables, which need checking is required if you write in `C`, but Python has better mechanisms for communicating errors. In Python this is not a good API.
Please raise exceptions and do proper processing with try/except.

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -250,7 +265,7 @@ def _get_full_name(name, version=None):
     if version is None or version == "":
         resp = name
     else:
-        resp = name + '==' + version
```
This could also move to a class after all.

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -271,15 +286,20 @@ def _get_packages(module, pip, chdir):
     '''Return whether or not package is installed.'''
+    if version is not None and version != "":
+        try:
+            vp = VersionPredicate("%s (%s)" % (name, _version_number_to_formula(version)))
```
Why use parens syntax?

HD650 replied:
VersionPredicate doesn't support `Django >=1.11.1` but only support `Django (>=1.11.1)`. They need brackets to help them separate name and version.

webknjaz replied:
I've added a comment for you to consider switching to pkg_resources

webknjaz reviewed (commented):
I'm not sure whether it's in the scope of this problem, but it would be nice to support `-c|--constraints` argument as well. Maybe even pass all packages without version, but also bypass version constraints via a separate argument.

Commit:
Add integration test for multi-pkg installing

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -222,19 +222,34 @@
 import re
 import sys
 import tempfile
+from distutils.versionpredicate import VersionPredicate
```
So I've tried a specifier, supported by pip and it turns out that distutils' parser don't support just a major version

```python-traceback
In [56]: vp = VersionPredicate('cherrypy(==4)')
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-56-b8d12ec3c098> in <module>()
----> 1 vp = VersionPredicate('cherrypy(==4)')

~/.pyenv/versions/3.6.5/lib/python3.6/distutils/versionpredicate.py in __init__(self, versionPredicateStr)
    114                 raise ValueError("expected parenthesized list: %r" % paren)
    115             str = match.groups()[0]
--> 116             self.pred = [splitUp(aPred) for aPred in str.split(",")]
    117             if not self.pred:
    118                 raise ValueError("empty parenthesized list in %r"

~/.pyenv/versions/3.6.5/lib/python3.6/distutils/versionpredicate.py in <listcomp>(.0)
    114                 raise ValueError("expected parenthesized list: %r" % paren)
    115             str = match.groups()[0]
--> 116             self.pred = [splitUp(aPred) for aPred in str.split(",")]
    117             if not self.pred:
    118                 raise ValueError("empty parenthesized list in %r"

~/.pyenv/versions/3.6.5/lib/python3.6/distutils/versionpredicate.py in splitUp(pred)
     24         raise ValueError("bad package restriction syntax: %r" % pred)
     25     comp, verStr = res.groups()
---> 26     return (comp, distutils.version.StrictVersion(verStr))
     27
     28 compmap = {"<": operator.lt, "<=": operator.le, "==": operator.eq,

~/.pyenv/versions/3.6.5/lib/python3.6/distutils/version.py in __init__(self, vstring)
     38     def __init__ (self, vstring=None):
     39         if vstring:
---> 40             self.parse(vstring)
     41
     42     def __repr__ (self):

~/.pyenv/versions/3.6.5/lib/python3.6/distutils/version.py in parse(self, vstring)
    135         match = self.version_re.match(vstring)
    136         if not match:
--> 137             raise ValueError("invalid version number '%s'" % vstring)
    138
    139         (major, minor, patch, prerelease, prerelease_num) = \

ValueError: invalid version number '4'

```

The one from `pkg_resources`, which is being used by pip, works though:
```python
from pkg_resources import Requirement
rq = Requirement('cherrypy=4.0')

In [58]: rq.specifier.contains('4')
Out[58]: True

In [59]: rq.specifier.contains('5')
Out[59]: False

In [60]: rq.specifier.contains('4.0.0')
Out[60]: True

In [61]: rq.specifier.contains('4.0.1')
Out[61]: False

In [62]: Requirement.parse('cherrypy>=4,<6').specifier.contains('5')
Out[62]: True

In [63]: Requirement.parse('cherrypy>=4,<6').specifier.contains('6')
Out[63]: False

```

@HD650 please switch to `pkg_resources.Requirement`

HD650 replied:
You are right, VersionPredicate only supports full version number but users may input some simplified version number.

Commit:
Improve code style

Commit:
Update test cases

Commit:
Support multi packages in one name string

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -222,10 +222,12 @@
 from ansible.module_utils.basic import AnsibleModule, is_executable
 from ansible.module_utils._text import to_native
 from ansible.module_utils.six import PY3
+from ansible.module_utils.common.collections import is_string
```
@HD650 do you still need this import?

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -235,6 +237,61 @@
+        Requirement(name)
+        return True
+    except RequirementParseError:
+        return False
```
please note that behavior for non-caught exceptions is to return `None`, so please add another `return False` in the end and maybe replace this one with `pass` or a docstrinig with the explanation.

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -235,6 +237,61 @@
+
+
+def _fetch_one_pkg(name):
+    stack = list()
```
Technically, it's not a stack data structure, so it's better to pick other name, which also reflects the purpose of it.
Something like `single_package` or `package_parts` might be better.

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -235,6 +237,61 @@
+    yield ",".join(stack)
+
+
+def _parses_name_input(module, name):
```
> parses

`parse`, because it's an action. But I'd go for smth like `get_packages_from_specs`.

webknjaz replied:
I'd also turn `name` into `names` because it's a list and variable name should reflect this detail.

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -235,6 +237,61 @@
+    """Parse name list to package list."""
+    package_list = list()
+    try:
+        for package_name in _fetch_one_pkg(name):
```
You don't need to create a `list()` and then mutate it. Instead, you can turn it into a list comprehension, making the whole function shorter and better readable:
```python
try:
    return [Package(name) for name in _recover_package_names(names)]
except Exception as e:  #  includes RequirementParseError, but also other errors, which might occur, and would need fail_json anyway
    module.fail_json(msg=str(e))
```

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -235,6 +237,61 @@
+        return False
+
+
+def _fetch_one_pkg(name):
```
You don't fetch anything here but reconstruct broken input. You return a sequence of data, just lazily. So name the function like it would return a list.
I think you could go for `_recover_package_names` here.

webknjaz replied:
And, of course, it accepts `names`, not one name.

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -235,6 +237,61 @@
+class Package(Requirement):
+    def __init__(self, name_string, version_string=None):
+        if version_string is not None:
+            if version_string[0].isdigit():
```
This would probably look cleaner:
```python
separator = '==' if version_string[0].isdigit() else ''
name_string = separator.join((name_string, version_string))
```

webknjaz replied:
Also, it might make sense to `str.lstrip()` `version_string` before checking the first char.

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -235,6 +237,61 @@
+        super(Package, self).__init__(name_string)
+
+    def have_version_specifier(self):
+        if len(self.specs) == 0:
```
Whole function body can be just `return bool(self.specs)`

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -235,6 +237,61 @@
+                name_string = "%s %s" % (name_string, version_string)
+        super(Package, self).__init__(name_string)
+
+    def have_version_specifier(self):
```
Please use `has` in the beginning.
Also, I feel like this should be a `@property`.

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -235,6 +237,61 @@
+            return False
+        return True
+
+    def get_name(self):
```
just have
```python
@property
def name(self):
    ...
```

webknjaz replied:
I'd call it `normalized_name` and would probably add some caching, so it would not be evaluated every time you access it.

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -235,6 +237,61 @@
+    def get_name(self):
+        return self.name.lower()
+
+    def satisfied_by(self, version_to_test):
```
Prefix it with `is_` to indicate a boolean check.

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -235,6 +237,61 @@
+    def satisfied_by(self, version_to_test):
+        return self.specifier.contains(version_to_test)
+
+    def install_command(self):
```
1. `@property`.
2. It's not an installation command, it's a version spec.
3. You don't need it, as it's stringified automatically during string interpolation, under 3.6.5:
```python
In [3]: from pkg_resources import Requirement

In [5]: rqc = Requirement('cherrypy')

In [11]: '%s ' % rqc
Out[11]: 'cherrypy '

```

2.6.9:
```python
$ python
Python 2.6.9 (unknown, Apr 10 2018, 17:32:50)
[GCC 7.3.0] on linux4
Type "help", "copyright", "credits" or "license" for more information.
>>> from pkg_resources import Requirement
/home/wk/.pyenv/versions/2.6.9/lib/python2.6/site-packages/pkg_resources.py:17: DeprecationWarning: the sets module is deprecated
  from sets import ImmutableSet
>>> rqc = Requirement('cherrypy')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: __init__() takes exactly 4 arguments (2 given)
>>> rqc = Requirement.parse('cherrypy')
>>> '%s ' % rqc
'cherrypy '
>>>
```

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -235,6 +237,61 @@
+                name_string = "%s ==%s" % (name_string, version_string)
+            else:
+                name_string = "%s %s" % (name_string, version_string)
+        super(Package, self).__init__(name_string)
```
Apparently, it will not work well under Python 2.6.9. It seems to be requiring you to call `Requirement.parse()` to construct an instance of this class. So we'll need to shuffle a couple of things around to make it work.

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -522,8 +583,8 @@ def main():
-            for pkg in name:
-                cmd += ' %s' % _get_full_name(pkg, version)
+            for pkg in packages:
+                cmd += ' \"%s\"' % pkg.install_command()
```
You don't need to escape double quotes inside of single quotes.

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -522,8 +583,8 @@ def main():
         if name:
-            for pkg in name:
-                cmd += ' %s' % _get_full_name(pkg, version)
+            for pkg in packages:
```
Also, you could go for a generator comprehension with extending the command:
```python
cmd += ''.join(' "%s"' % dist for dist in distributions)
```

webknjaz replied:
P.S. In Python, we call installable bundles "distributions", we call `.py` files "modules" and we call folders, which contain `__init__.py` and other `.py` files "packages".
Let's not confuse the term.

Commit:
Fix sanity issues and improve code style

Commit:
Fix error when install by url link

Commit:
Handle corner case when version is empty

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -223,6 +223,12 @@
 
+NO_SETUPTOOLS = False
+try:
+    from pkg_resources import Requirement, RequirementParseError
```
Better do `HAS_SETUPTOOLS` here and in except branch.

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -235,6 +241,72 @@
                              'pip': 'import pkg_resources; print(pkg_resources.get_distribution("pip").version)'}
 
 
+class Distribution(object):
```
You don't need to inherit from `object`, because `__metaclass__ = type` above already turns on new-style classes for everything by default.

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -235,6 +241,72 @@
+                name_string = separator.join((name_string, version_string))
+            self._requirement = Requirement.parse(name_string)
+            self._distribution_name = self._requirement.name.lower()
+        else:
```
If you could swap if and else parts, you'd avoid `not` negation in startswith check and also could omit `else` and dedent everything what's in if part now.

HD650 replied:
I didn't get, how to omit `else`? We should only execute one of the two code blocks.

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -235,6 +241,72 @@
+    def distribution_name(self):
+        return self._distribution_name
+
+    @distribution_name.setter
```
Since you don't use setter you probably don't need it.

HD650 replied:
in `__init__` I need a `self.__distribution_name = self._requirement.name.lower()`

webknjaz replied:
this setter runs when you to do `self.distribution_name = ...`

webknjaz replied:
Since you assign to underlying private/protected variable, you don't need this public interface for someone external to modify it.

HD650 replied:
ah, that error happened when I didn't rename the `distribution_name` to `_distribution_name`. You are right

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -271,15 +335,15 @@ def _get_packages(module, pip, chdir):
             continue
 
-        if pkg_name == name and (version is None or version == pkg_version):
+        if (pkg_name.lower() == req.distribution_name) and (req.is_satisfied_by(pkg_version)):
```
In Python, we avoid adding unnecessary brackets.

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -235,6 +241,72 @@
+    def is_satisfied_by(self, version_to_test):
+        if self._plain_distribution:
+            return self._requirement.specifier.contains(version_to_test)
+        return True
```
This is tricky. It should probably return false because there's no way of knowing whether it's the same version, so we'll need to invoke pip to reinstall it as it does when you run it manually

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -235,6 +241,72 @@
+
+
+def _is_valid_distribution_name(name):
+    if name.startswith(('>=', '<=', '!=', '==', '>', '<')):
```
It could be just plain `return not ...startswith...`

webknjaz replied:
Also, it seems that you don't strip spaces, so if there will be leading spaces it would not match, for example.

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -235,6 +241,72 @@
+
+
+def _recover_distribution_name(names):
+    distribution_parts = list()
```
please use list literal `[ ]`

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -235,6 +241,72 @@
+
+def _get_distributions_from_names(module, names):
+    """Parse name list to package list."""
+    package_list = list()
```
use list literal

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -456,6 +520,9 @@ def main():
     umask = module.params['umask']
     env = module.params['virtualenv']
 
+    if NO_SETUPTOOLS:
```
move this before var assignments to fail early.

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -503,12 +570,24 @@ def main():
+            if version is not None:
+                if len(distributions) > 1:
+                    module.fail_json(msg="Version arg is not available for installing multi-packages.")
+                else:
```
I think you could go for `elif:` here.

webknjaz replied:
or entirely omit `else:` since it would fail in case of `if` part.

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -522,8 +601,8 @@ def main():
-            for pkg in name:
-                cmd += ' %s' % _get_full_name(pkg, version)
+            for distribution in distributions:
+                cmd += ' "%s"' % distribution
```
```python
cmd += ''.join(' "%s"' % d for d in distributions)
```

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -503,12 +570,24 @@ def main():
+                    module.fail_json(msg="Version arg is not available for installing multi-packages.")
+                else:
+                    if distributions[0].has_version_specifier:
+                        module.fail_json(msg="Version arg is not available if version info is provided in name.")
```
I'd rephrase this into `'version' argument conflicts with version specifier provided along with a package name. Please keep a version specifier, but remove the 'version' argument.`

Commit:
Fix import error and improve code style

Commit:
Compatible with old version setuptools

mattclay commented:
Traceback in integration tests:

```
Traceback (most recent call last):
  File "/tmp/ansible_g5xgA6/ansible_module_pip.py", line 677, in <module>
    main()
  File "/tmp/ansible_g5xgA6/ansible_module_pip.py", line 638, in main
    is_present = _is_present(module, distribution, pkg_list, pkg_cmd)
  File "/tmp/ansible_g5xgA6/ansible_module_pip.py", line 345, in _is_present
    if pkg_name.lower() == req.distribution_name and req.is_satisfied_by(pkg_version):
  File "/tmp/ansible_g5xgA6/ansible_module_pip.py", line 274, in is_satisfied_by
    return self._requirement.specifier.contains(version_to_test)
AttributeError: Requirement instance has no attribute 'specifier'
```

Commit:
Try to fix

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -223,6 +223,12 @@
 import tempfile
 
+try:
+    HAS_SETUPTOOLS = True
```
Please set it *after* import succeeded.

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -235,6 +241,73 @@
+    def is_satisfied_by(self, version_to_test, module):
+        if self._plain_distribution and hasattr(self._requirement, 'specifier'):
+            return self._requirement.specifier.contains(version_to_test)
+        else:
```
After further thinking, I've recalled that URL-like deps can actually hold metadata through fragment, like: `#egg=package-1.0.0`.
Ref: https://pip.pypa.io/en/stable/reference/pip_install/#vcs-support
I'm not sure what is a proper strategy of reacting to this case though.

webknjaz replied:
It looks like we could compare if version points to a commit hash or a tag, which are effectively immutable, unlike a branch name.

webknjaz replied:
On the other hand, only pip can detect that, since it downloads dists, unless we do `pip download` ourselves and check what's inside.

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -235,6 +241,73 @@
+
+    def is_satisfied_by(self, version_to_test, module):
+        if self._plain_distribution and hasattr(self._requirement, 'specifier'):
+            return self._requirement.specifier.contains(version_to_test)
```
What if version_to_test doesn't contain a valid specifier, being installed from git url or with `--editable` mode. Can we use `pip list` in addition to `pip freeze` to grab more info about installed dists?

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -235,6 +241,73 @@
+        if self._plain_distribution and hasattr(self._requirement, 'specifier'):
+            return self._requirement.specifier.contains(version_to_test)
+        else:
+            module.warn(str(self._requirement.__dict__))
```
By the way, you don't need `else:`, since this will be evaluated anyway.

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -235,6 +241,73 @@
                              'pip': 'import pkg_resources; print(pkg_resources.get_distribution("pip").version)'}
 
 
+class Distribution:
```
Could you, please, write PEP257-compatible docstrings to all public interfaces you add?

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -235,6 +241,73 @@
+
+
+def _is_valid_distribution_name(name):
+    name = name.lstrip()
```
Why don't you do this right in the `return` line?
If you think it's going to be too long, be aware that you can break the line by wrapping it with braces:
```python
def _is_valid_distribution_name(name):
    return (
        not name.
        lstrip().
        startswith(('>=', '<=', '!=', '==', '>', '<'))
    )
```
(you can put `not` outside of brackets if you like it more this way)

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -235,6 +241,73 @@
+    for name in names:
+        if _is_valid_distribution_name(name):
+            if len(distribution_parts) > 0:
+                yield ",".join(distribution_parts)
```
I'd probably add `distribution_parts = []` after yield inside of `if` block and then you'd only need `distribution_parts.append(name)` outside of if-block without a need to have diverse code or `else`-block.

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -235,6 +241,73 @@
+
+def _get_distributions_from_names(module, names):
+    """Parse name list to package list."""
+    package_list = []
```
This is unnecessary and unused variable.

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -271,15 +336,15 @@ def _get_packages(module, pip, chdir):
     for pkg in installed_pkgs:
         if '==' in pkg:
             pkg_name, pkg_version = pkg.split('==')
         else:
```
Hm... Does it skip `--editable` dists this way? Can/should `pip list` improve this experience?

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -235,6 +241,73 @@
+    def distribution_name(self, new_name):
+        self._distribution_name = new_name
+
+    def is_satisfied_by(self, version_to_test, module):
```
Look what I've found:

```python
$ ipython
Python 3.7.0 (default, Jun 28 2018, 11:18:11)
Type 'copyright', 'credits' or 'license' for more information
IPython 6.4.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import pkg_resources

In [10]: ipdst = pkg_resources.get_distribution('ipython')

In [11]: ipdst
Out[11]: ipython 6.4.0 (/home/wk/.pyenv/versions/3.7.0/envs/ansible-py3.7.0-pyenv-venv/lib/python3.7/site-packages)

In [15]: ipdst.as_requirement()
Out[15]: Requirement.parse('ipython==6.4.0')

```

This probably can completely eliminate need for doing `pip freeze` and parsing it!

@HD650 please investigate this.

webknjaz replied:
```python
$ python2.6
Python 2.6.9 (unknown, Apr 10 2018, 17:32:50)
[GCC 7.3.0] on linux4
Type "help", "copyright", "credits" or "license" for more information.
>>> import pkg_resources
/home/wk/.pyenv/versions/2.6.9/lib/python2.6/site-packages/pkg_resources.py:17: DeprecationWarning: the sets module is deprecated
  from sets import ImmutableSet
>>> stdst = pkg_resources.get_distribution('setuptools')
>>> stdst
setuptools 0.6b4dev-r0 (/home/wk/.pyenv/versions/2.6.9/lib/python2.6/site-packages)
>>> stdst.as_requirement()
Requirement.parse('setuptools==0.6b4dev-r0')
>>> stdst.as_requirement().specs
[('==', '0.6b4dev-r0')]
```

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -503,12 +571,24 @@ def main():
+                if distributions[0].has_version_specifier:
+                    module.fail_json(msg="'version' argument conflicts with version specifier provided along with a package name.\
+                                     Please keep a version specifier, but remove the 'version' argument.")
+                else:
```
Could dedent here if `else:` removed.

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -503,12 +571,24 @@ def main():
+                if len(distributions) > 1:
+                    module.fail_json(msg="'version' is not available for installing multi-packages.")
+                if distributions[0].has_version_specifier:
+                    module.fail_json(msg="'version' argument conflicts with version specifier provided along with a package name.\
```
You don't need to escape EOL. Just have two strings and Python will automatically know that it's in fact one string:

this
```python
(
"series "
"of"
" words"
)
```
equals
```python
"series of words"
```

webknjaz replied:
(but in case of function arg, you don't need additional braces)

webknjaz replied:
```python
   func("one "
        "string",
        "another" " string")
```

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -503,12 +571,24 @@ def main():
+                    module.fail_json(msg="'version' argument conflicts with version specifier provided along with a package name.\
+                                     Please keep a version specifier, but remove the 'version' argument.")
+                else:
+                    distributions[0] = Distribution(distributions[0].distribution_name, version)
```
Why don't you assign string here? Your further loops assume it's a string and convert it for you anyway.

HD650 replied:
There is a more further loop in check_mode that need the `distributions` to be instances of `Distribution`.

webknjaz commented on test/integration/targets/pip/vars/main.yml:
```diff
@@ -2,6 +2,12 @@ pip_test_package: sampleproject
   - sampleproject
   - decorator
+pip_test_pkg_ver:
+  - sampleproject<=100,!=9.0.0,>=0.0.1
```
You might want to test with spaces in different places.

HD650 replied:
I will fix these with some new test cases added together.

webknjaz replied:
so?

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -223,6 +223,12 @@
+    HAS_SETUPTOOLS = True
+    from pkg_resources import Requirement
+except ImportError:
+    HAS_SETUPTOOLS = True
```
`False`

Commit:
Add fallback if setuptools is too old

Commit:
Fix package name error in centos6

Commit:
Refactor and improve code style

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -235,6 +244,95 @@
 
+class Distribution:
+
+    """Wrapper class for pkg_resources.Requirement.
```
It's an implementation detail. The docstring title should tell the reader class' purpose rather than which techniques you used. I'd go for `Python distribution package metadata wrapper.`

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -235,6 +244,95 @@
+
+    A wrapper class for Requirement, provide usefual API to parse package name and version specifier,
+    do fallback if the installed setuptools version doesn't support some features.
+
```
Normally, it's a style violation to have an empty string at the end of docstring.

HD650 replied:
I may have read a fake document... I will fix this.

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -235,6 +244,95 @@
+
+    def is_satisfied_by(self, version_to_test):
+        if self._plain_distribution:
+            if hasattr(self._requirement, 'specifier'):
```
I'd go for try/except style

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -235,6 +244,95 @@
+                op_dict = {">=": operator.ge, "<=": operator.le, ">": operator.gt,
+                           "<": operator.lt, "==": operator.eq, "!=": operator.ne}
+                # old setuptools has no specifier, do fallback
+                for (op, ver) in self._requirement.specs:
```
you don't need parens here

webknjaz replied:
This could be:
```python
return all(
    op_dict[op](version_to_test, LooseVersion(ver))
    for op, ver in self._requirement.specs
)
```

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -235,6 +244,95 @@
+                version_to_test = LooseVersion(version_to_test)
+                op_dict = {">=": operator.ge, "<=": operator.le, ">": operator.gt,
+                           "<": operator.lt, "==": operator.eq, "!=": operator.ne}
+                # old setuptools has no specifier, do fallback
```
I'd do:
```python
from functools import partial
...
def is_satisfied_by(current_version, constraint_predicate):
    op, ver = constraint_predicate
    return op_dict[op](current_version, LooseVersion(ver))

satisfies_target_version = partial(is_satisfied_by, current_version=LooseVersion(version_to_test))

return all(
    satisfies_target_version(spec)
    for spec in self._requirement.specs
)
# or return all(map(satisfies_target_version, self._requirement.specs))
```

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -235,6 +244,95 @@
+
+def _is_valid_distribution_name(name):
+    """Test whether a name is distribution name or version specifier."""
+    return not name.lstrip().startswith(('>=', '<=', '!=', '==', '>', '<'))
```
`'>=', '<=', '!=', '==', '>', '<'` is duplicated with that operator map. you could do global map and use `that_map.keys()`

Commit:
Supplement docstring and improve code style

HD650 commented:
@abadger @robinro Now the code runs well, could you review again and give some advice? Thanks!

webknjaz commented:
@HD650 please fix a merge conflict.

webknjaz commented:
@dagwieers instead of deleting comments you can now "Hide" them, which effectively collapses messages and is better if someone wants to revise them in the future.

HD650 commented:
@webknjaz Is there a conflict? It's green in my browser.

webknjaz committed:
Fix Distribution docstring

webknjaz committed:
Dedent code in Distribution initializer

Also fix bug with `self._requirement` not being initialized in one code branch

webknjaz committed:
Make is_satisfied_by more pythonic

webknjaz committed:
Turn tangled for-loop into a comprehension with all predicate

webknjaz committed:
Simplify recovered dists generator

webknjaz committed:
Turn dict constructor into a literal

webknjaz committed:
Unbracket returning tuple

webknjaz committed:
Improve failure wording

webknjaz committed:
Pythonify cmd construction

webknjaz commented on test/integration/targets/pip/tasks/pip.yml:
```diff
@@ -273,3 +273,146 @@
+
+- name: assert no changes ocurred
+  assert:
+    that: "version2.changed == False"
```
use `not version2.changed` instead

webknjaz commented on test/integration/targets/pip/vars/main.yml:
```diff
@@ -2,6 +2,12 @@ pip_test_package: sampleproject
 pip_test_packages:
   - sampleproject
   - decorator
+pip_test_pkg_ver:
```
I would also check installs from Git + `--editable` installs.

HD650 replied:
There is already one: https://github.com/ansible/ansible/blob/devel/test/integration/targets/pip/tasks/pip.yml#L145

Commit:
Fix the parse failure to url input

Package string parsing should only happen when pack string
is a name but not a url.

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -234,6 +260,94 @@
+            version_string = version_string.lstrip()
+            separator = '==' if version_string[0].isdigit() else ' '
+            name_string = separator.join((name_string, version_string))
+        self._requirement = Requirement.parse(name_string)
```
if this method terminates in line 278 there will be no `self._requirement` in class instance, which would break things.

HD650 replied:
To access `_requirement` it will first check `_plain_distribution`, so it's ok.

webknjaz replied:
It's inconsistency, so it's not ok.

abadger replied:
I agree with @webknjaz, set self._requirement = None if you have to... just so that it is set in both code paths.

Commit:
Modify some wording

robinro commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -24,9 +24,11 @@
     description:
       - The name of a Python library to install or the url of the remote package.
       - As of 2.2 you can supply a list of names.
+      - As of 2.7, you can supply a list of names with version specifiers.
```
Please don't repeat the previous sentence, better:
`- This can be a list (since 2.2) and contain version specifiers (since 2.7).`

robinro commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -24,9 +24,11 @@
   version:
     description:
       - The version number to install of the Python library specified in the I(name) parameter.
+      - As of 2.7, you can supply version specifiers.
```
I would not mention version specified here, but recommend to use them only in the name. "There should be one-- and preferably only one --obvious way to do it."

robinro commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -125,6 +127,21 @@
     name: bottle
```
As stated above I would recommend to have one recommended way to specify the version, i.e change this to `name: bottle==0.11`

robinro commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -125,6 +127,21 @@
+# Install (bottle) python package with version specifiers (only in 2.7)
+- pip:
+    name: bottle
+    version: '>0.10,<0.20,!=0.11'
```
As above.

robinro commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -125,6 +127,21 @@
+      - 'django>1.11.0,<1.12.0'
+      - 'bottle>0.10,<0.20,!=0.11'
+
+# 2.7 also support multi packages in one name string
```
Please remove this example, since I would consider this usage as not recommended.

robinro commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -125,6 +127,21 @@
     name: bottle
     version: 0.11
 
+# Install (bottle) python package with version specifiers (only in 2.7)
```
There is no need to specify the version here.

robinro commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -125,6 +127,21 @@
+    name: bottle
+    version: '>0.10,<0.20,!=0.11'
+
+# Install multi python packages with version specifiers (only in 2.7)
```
As above.

robinro commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -234,6 +260,94 @@
+        return False
+
+    @property
+    def distribution_name(self):
```
This is unnecessary, you can write/access `self.distribution_name` directly and save one indirection.

robinro commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -234,6 +260,94 @@
+    return not name.lstrip().startswith(tuple(op_dict.keys()))
+
+
+def _recover_distribution_name(names):
```
As far as I understand this is only necessary to support stuff like `djang>1.11.0,<1.12.0,bottle>0.10,<0.20,!=0.11`.
Why should we support this? Using a list is to me the obviously better API and support this structure adds a lot of (to me unnecessary) complexity.

abadger replied:
Cross-compability, backwards compatibility, command line compatibility, ease of discovery.

* Other packaging modules support comma separated package entries.
* The pip module itself supports comma separated package entries
* pip supports single string entries as valid packages
* comma separated list entries are the natural way to write lists on the commandline.
* error messages if people were to violate these rules (for instance: ```ansible localhost -m pip -a 'names=django>=1.11.0,<1.12.0'``` wonn't point to the actual solution unless we implement this sort of detection for creating the error message... in which case we might as well do the right thing.

robinro commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -234,6 +260,94 @@
+    try:
+        return [Distribution(name) for name in _recover_distribution_name(names)]
+    except ValueError as e:
+        module.fail_json(msg=str(e))
```
Can you give an example for the error message coming from this? As stated above I would rather remove support for construtcs, which can raise exceptions here.

HD650 replied:
This error comes from the `__init__` of `Distribution`, since the user may input some invalid package name like `djan--go`. The `Requirement.parse(name_string)` will find those errors and raise an error.

robinro commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -503,12 +612,28 @@ def main():
+            if version is not None:
+                if len(distributions) > 1:
+                    module.fail_json(
+                        msg="'version' argument is ambiguous when installing multiple package distributions. "
```
I would call them `packages` instead of `distributions`.

webknjaz replied:
No, the correct name is `distributions`. In Python, packages are folders with python modules (`.py` files) and `__init__.py`, so that would be confusing terminology.
OTOH `package distributions` is a known compromise term for those doesn't really know about how python packaging works and its terminology.

robinro replied:
> It is common in the Python community to refer to a distribution using the term “package”. Using the term “distribution” is often not preferred, because it can easily be confused with a Linux distribution, or another larger software distribution like Python itself.

https://packaging.python.org/tutorials/installing-packages/

robinro replied:
"package" is more common in documentation and examples.

webknjaz replied:
Well, it was used to be known worse a while back. They've started working on that tutorial not long ago. Old users don't rush to reread those. That's why I prefer combo compromise to be as explicit as possible. Python packaging is in a weird state for a long time now. I don't want to add any more confusion.

abadger replied:
FWIW, +1 to package.

Distribution is only applicable to python, making it harder for people outside of that ecosystem to work on this module.

abadger replied:
Actually, even more than +1 to changing this to packages because this is in the error messages which we return to the user, not just about variable naming.  Everyone will understand what packages means in this context but sysadmins who just need to translate a script to setup the production environment their dev team gave them won't understand "distribution".

robinro commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -503,12 +612,28 @@ def main():
+                if len(distributions) > 1:
+                    module.fail_json(
+                        msg="'version' argument is ambiguous when installing multiple package distributions. "
+                            "Please spefify version restrictions next to each distribution in 'name' argument."
```
typo: `specify`

robinro commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -503,12 +612,28 @@ def main():
+                if distributions[0].has_version_specifier:
+                    module.fail_json(
+                        msg="'version' argument conflicts with version specifier provided along with a package name. "
+                            "Please keep a version specifier, but remove the 'version' argument."
```
Better: **The** 'version' argument conflicts with any version specifier provided along with a package name. 
Please keep **the** version specifier, but remove the 'version' argument.

robinro reviewed (changes requested):
Thanks @HD650 for the effort you already put into this PR. To me it looks like it passed most of the way to getting merged.
I'm a little worried about the added complexity and would therefore recommend to not support multiple packages in a line with version specifiers.

Commit:
Adjust documents and examples

Commit:
Remove unnecessary property

Commit:
Make parsing logic easier to read

HD650 commented:
@robinro I have discussed the multi-packages in one line feature with @abadger, we think we should keep this feature. And it's not much more complex than inputting a list of packages, just need some lines of code to reconstruct the input.

robinro reviewed (approved):
Looks good overall now. I'm still not happy about adding the new input format, but everyone else seems to like it, so from my side this is ready to be merged.

HD650 commented:
@robinro Thanks, I will add more tests to make sure this PR doesn't introduce any bugs.

HD650 commented:
@robinro After some tests I kinda agree with you... This pr is too big and complex and it's hard to test, I will remove the support of one-line-multi-packages in this PR and do that in another PR. This PR will only have basic support for version specifiers.

Commit:
Tolerate all combinations of input

Convert list of name, list of part of name or list of
name and part of name to falt list. This can make the
program tolerate all kinds of input

Commit:
Add more strict test case

Commit:
Make parsing error easy to read

Commit:
Test space in version specifier

Commit:
Fix sanity test error

abadger commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -234,6 +252,87 @@
+
+def _is_vcs_url(name):
+    """Test whether a name is a vcs url or not."""
+    return re.match(r'(svn|git|hg|bzr)\+', name)
```
I usually pre-compile regexes into a global variable so that it only has to be done once.

```
_VCS_RE = re.compile(r'(svn|git|hg|bzr)\+')

[...]
def _is_vcs_url(name):
    """Test whether a name is a vcs url or not."""
    return re.match(_VCS_RE, name)
```

abadger replied:
There's a comment where this was extracted from the code which should probably be moved to this docstring:

> VCS includes those beginning with "svn+", "git+", "hg+" or "bzr+"

abadger commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -234,6 +252,87 @@
+           "<": operator.lt, "==": operator.eq, "!=": operator.ne}
+
+
+class Distribution:
```
Style note: toplevel functions should be placed above class definitions)

abadger commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -234,6 +252,87 @@
+
+
+def _recover_distribution_name(names):
+    """Recover distribution names as list from user's raw input."""
```
Whenever you're doing more than simle parsing (simple parsing is usually, one command, like ```variable.split(",")```) you should give some examples of what the code is expecting as input and what it will give as output.

abadger commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -234,6 +252,87 @@
+def _recover_distribution_name(names):
+    """Recover distribution names as list from user's raw input."""
+    # rebuild input name to a flat list so we can tolerate any combination of input
+    names = [one_part for one_line in names for one_part in one_line.split(",")]
```
Nesting generator expressions and list comprehensions like this is bad style.  It has the same problems as a  run-on sentence in natural languages (Making it hard for other people to read and keep the entirety of the clause in their memory.  Easy to misinterpret the meaning because of misreading one small piece of the grammar).

In most cases, using a ```for``` loop with indentation for at least one of the loops is better.

If I understand this correctly,  You are trying to take input of this form:
```
names = ['one >1.0,<2.0', 'two', '>3.0', '<4.0']
```
and flatten it so that it is in this form:
```
names = ['one', '>1.0', '<2.0', 'two', '>3.0', '<4.0']
```
?

If so, it's not quite right as it will currently return 
```
 ['one >1.0', '<2.0', 'two', '>3.0', '<4.0']
```

HD650 replied:
Yes we will get `one >1.0` unsplited, but pip and `Distribution` can handle that format. This line is intend to handle this:
If user input 
```
name:
  - django>1.11.1,<1.12.1,ipaddress
  - simpleproject==1.1.0
```
We will get `['django>1.11.1,<1.12.1,ipaddress', 'simpleproject==1.1.0']`, but we want django separat from ipaddress, so here I split all elements with `,` and then concate them again.

abadger commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -234,6 +252,87 @@
+    distribution_names = []
+    for name in names:
+        if _is_valid_distribution_name(name):
+            if len(name_parts) > 0:
```
Here you can simply write:
```
if name_parts:
```

Python considers empty containers to be False-y.

abadger commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -234,6 +252,87 @@
+            name_parts = []
+        name_parts.append(name)
+    distribution_names.append(",".join(name_parts))
+    return distribution_names
```
Currently with this input:
```
names = ['one >1.0,<2.0', 'two', '>3.0', '<4.0']
```
You would get this output:
```
distribution_names = ['one >1.0,<2.0', 'two,>3.0,<4.0']
```

Note that one does not have a comma before the first version while two does.

HD650 replied:
This will happen if user input:
```
name: one >1.0,<2.0, two,>3.0,<4.0 #note the comma between two and >3.0
```
or
```
name:
  - one >1.0,<2.0
  - two
  - >3.0
  - <4.0
```
and those are invalid for `pip`, so I think we could not handle this case.

abadger commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -234,6 +252,87 @@
+    return not name.lstrip().startswith(tuple(op_dict.keys()))
+
+
+def _recover_distribution_name(names):
```
Functions which are side effect free and don't deal with external information like this one are excellent choices to unittest.  You can give the function a wide range of potential inputs and check that they match up with the expected outputs far cheaper than you can with intergration tests.

abadger commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -521,11 +639,11 @@ def main():
+            cmd += ' '.join('"%s"' % d for d in distributions)
         elif requirements:
-            cmd += ' -r %s' % requirements
+            cmd += '-r "%s"' % requirements
```
I think you need to change cmd into a list and then pass that into module.run_command instead of trying to quote arguments here.  module.run_command() doesn't use a shell unless use_unsafe_shell=True is passed in.

abadger replied:
https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/basic.py#L2714

HD650 replied:
Make sense, but open another pr for this is better since all operations on `cmd` in pip module use pattern like `cmd = xxx + xxx`, it needs many changes at many places, and it's not relevant to version specifier problem.

Commit:
Improve code style and supplement docstring

Commit:
Rename Distribution to Package

abadger commented:
That's what I thought when I started to write my connection he but part way
through I realized it had to be fixed now.  Quotes are handled by Shell so
the package names and versions probably aren't doing the right thing.

On Tue, Jul 24, 2018, 11:34 AM Zhikang Zhang <notifications@github.com>
wrote:

> *@HD650* commented on this pull request.
> ------------------------------
>
> In lib/ansible/modules/packaging/language/pip.py
> <https://github.com/ansible/ansible/pull/41792#discussion_r204866091>:
>
> >          elif requirements:
> -            cmd += ' -r %s' % requirements
> +            cmd += '-r "%s"' % requirements
>
> Make sense, but open another pr for this is better since all operations on
> cmd in pip module use pattern like cmd = xxx + xxx, it needs many changes
> at many places, and it's not relevant to version specifier problem.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ansible/ansible/pull/41792#discussion_r204866091>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAMxWriP0XMXJsICuaQMa1ZwAsV-uEMSks5uJ2ijgaJpZM4UyHwi>
> .
>

Commit:
Add a unit test for recover_package_name func

Commit:
Fix pytest error

Commit:
Change cmd from string to list

run_command may use shell=True to run a string command
which is dangerous, so change cmd from string to list

Commit:
Add support to ~= and improve code style

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -234,6 +252,53 @@
+           "<": operator.lt, "==": operator.eq, "!=": operator.ne, "~=": operator.ge}
+
+
+def _is_vcs_url(name):
```
Remark: with regexp factored out this could become just
```python
from functools import partial
...
_is_vcs_url = partial(re.match, _VCS_RE)
"""Test whether a name is package name or version specifier."""
```

(But I bet Toshio would blame it for being unreadable, so don't do it :smile:)

abadger replied:
I wouldn't write a function which was itself a single function call ;-)

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -234,6 +252,53 @@
+
+
+def _is_valid_package_name(name):
+    """Test whether a name is package name or version specifier."""
```
Let native speakers correct me if I'm wrong, but I think `a` article should be moved to before `package` and before `version`.

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -234,6 +252,53 @@
+_VCS_RE = re.compile(r'(svn|git|hg|bzr)\+')
+
+op_dict = {">=": operator.ge, "<=": operator.le, ">": operator.gt,
+           "<": operator.lt, "==": operator.eq, "!=": operator.ne, "~=": operator.ge}
```
> "~="

I've never heard of it. What is the source of such an assumption?

HD650 replied:
I saw it here https://www.python.org/dev/peps/pep-0440/#version-specifiers, and it's supported by `pip`.

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -234,6 +252,53 @@
+    return re.match(_VCS_RE, name)
+
+
+def _is_valid_package_name(name):
```
This function checks whether input does not start with an operator, not that it's a package name at all. Let's invent a better name.

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -234,6 +252,53 @@
+    # rebuild input name to a flat list so we can tolerate any combination of input
+    tmp = []
+    for one_line in names:
+        tmp = tmp + one_line.split(",")
```
`list`s have `.append()` method. You probably don't want to re-allocate a new instance on each iteration of the loop.

HD650 replied:
I want the original name-specifier-mixed list like
`['django>1.11.1,ipaddress', 'sampleproject']`
become totatly deassemble like:
`['django','>1.11.1','ipaddress','sampleproject']`

abadger replied:
Lists also have .extend() which might be what you need here

abadger replied:
Lists also have .extend() which might be what you need here

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -234,6 +252,53 @@
+    tmp = []
+    for one_line in names:
+        tmp = tmp + one_line.split(",")
+    names = tmp
```
This whole block could've been done with a nested list comprehension without any need for temporary variable introduction:
```python
names = [
    name
    for name_list in names
        for name in name_list.split(',')
]
```

Alternative FP approach would look like:
```python
from itertools import partial, reduce
...
names = reduce(
    lambda l, e: l + e,
    map(
        partial(str.split, sep=','),
        names
    ),
    [],
)
```
(but Pythonistas don't like such style in general)

HD650 replied:
Toshio thinks it's better to use a for loop since it's easy to read so I changed to for loop.

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -234,6 +252,53 @@
+    for name in names:
+        if _is_valid_package_name(name):
+            if name_parts:
+                package_names.append(",".join(name_parts))
```
I liked generator approach more: it's more readable in terms of separating a distinct logical step from a higher-level flow.

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -417,12 +474,64 @@ def setup_virtualenv(module, env, chdir, out, err):
     return out, err
 
 
+class Package:
```
Maybe `DistributionPackage`?

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -417,12 +474,64 @@ def setup_virtualenv(module, env, chdir, out, err):
+        if version_string:
+            version_string = version_string.lstrip()
+            separator = '==' if version_string[0].isdigit() else ' '
+            name_string = separator.join((name_string, version_string))
```
Should we also update `self.package_name`?

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -447,6 +556,9 @@ def main():
     )
 
+    if not HAS_SETUPTOOLS:
+        module.fail_json(msg="No setuptools found in remote host, please install that first.")
```
`that` -> `it`

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -505,10 +617,33 @@ def main():
         if name:
             for pkg in name:
-                if bool(pkg and re.match(r'(svn|git|hg|bzr)\+', pkg)):
+                if bool(pkg and _is_vcs_url(pkg)):
```
`bool()` wrapper is unnecessary here. The expression is being cast to `bool` anyway.

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -505,10 +617,33 @@ def main():
 
+            # convert raw input package names to Package instances
+            try:
+                packages = []
```
I'd prefer a list comprehension:
```python
packages = [Package(dist) for dist in _recover_package_name(name)]
```

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -519,13 +654,13 @@ def main():
         if name:
-            for pkg in name:
-                cmd += ' %s' % _get_full_name(pkg, version)
+            for p in packages:
```
I'd prefer
```python
# a generator expression:
cmd.extend(str(p) for p in packages)
# or:
cmd.extend(map(str, packages))
```

abadger replied:
A generator expression is fine here.  Just remember to use to_native instead of str().

```
cmd.extend(to_native(p, errors='surrogate_or_strict') for p in packages)
```

abadger replied:
A generator expression is fine here.  Just remember to use to_native instead of str().

```
cmd.extend(to_native(p, errors='surrogate_or_strict') for p in packages)
```

webknjaz commented on test/integration/targets/pip/tasks/pip.yml:
```diff
@@ -273,3 +273,152 @@
+
+- name: assert no changes
+  assert:
+    that: "version3.changed == False"
```
The correct way is to do `not version3.changed`. Avoid equivalence comparison with bool.

webknjaz commented on test/integration/targets/pip/tasks/pip.yml:
```diff
@@ -273,3 +273,152 @@
+- name: test module can parse the combination of multi-packages one line and git url
+  pip:
+    name:
+      - git+https://github.com/dvarrazzo/pyiso8601#egg=pyiso8601
```
Do you parse metadata from URL fragment, like `#egg=name-version`?

Commit:
Small code style changes

abadger commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -519,13 +652,12 @@ def main():
         elif requirements:
-            cmd += ' -r %s' % requirements
+            cmd.extend(['-r', requirements])
         else:
```
Don't use str().  Use to_native (in module code) instead.  (Use to_text in controller code).

I left another comment with how to use to_native

webknjaz replied:
where's str? did you mean the implicit type of `cmd` contents?

Commit:
Use to_native and fix sanity error

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -519,13 +652,12 @@ def main():
         if name:
-            for pkg in name:
-                cmd += ' %s' % _get_full_name(pkg, version)
+            cmd.extend(str(p) for p in packages)
```
I believe @abadger meant this place. So you basically want to use `to_native()` instead.

Commit:
Adjust the try block

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -417,12 +474,67 @@ def setup_virtualenv(module, env, chdir, out, err):
+        try:
+            self._requirement = Requirement.parse(name_string)
+        except ValueError as e:
+            raise ValueError("Can not parse package name '%s', error: '%s'" % (name_string, to_native(e)))
```
`six.raise_from(ValueError("Can not parse package name '%s', error: '%s'" % (name_string, to_native(e))), e)`

robinro replied:
Is this still necessary? There should be test coverage for this line by adding some broken requirements specification.

Commit:
Remove the invalid package name error

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -486,23 +486,21 @@ def __init__(self, name_string, version_string=None):
+                self.package_name = "setuptools"
+            else:
+                self.package_name = self._requirement.project_name
+            self._plain_package = True
```
This attribute will now exist only in selected code paths, which is bad.

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -486,23 +486,21 @@ def __init__(self, name_string, version_string=None):
+                self.package_name = self._requirement.project_name
+            self._plain_package = True
         except ValueError as e:
-            raise ValueError("Can not parse package name '%s', error: '%s'" % (name_string, to_native(e)))
```
What is the reason for removing error handling?

HD650 replied:
Like @robinro said, it's not necessary, `pip` can handle the invalid package name.
Another thing is, users can input a file name or a directory name to `name`, like`file:\\xxxx`, 
 `django.tar.gz` or `./django/`, we should only parse the plain package name and leave all other types of name to `pip`, let `pip` decide whether a file or directory is valid to install.

webknjaz replied:
ok

webknjaz reviewed (changes requested):
I disagree with dropping proper error handling.

jctanner commented:
bot_skip

webknjaz commented on lib/ansible/modules/packaging/language/pip.py:
```diff
@@ -529,10 +529,10 @@ def __str__(self):
-        absent=['uninstall', '-y'],
-        latest=['install', '-U'],
-        forcereinstall=['install', '-U', '--force-reinstall'],
+        present='install',
```
Why did you revert this to string?

HD650 replied:
When user input extra option like `--user --root xxxxxx`, the list type cannot handle it well.

webknjaz replied:
well, it can always be str.split

Commit:
Lexically split extra arguments
